src/analyses/dependence_graph.cpp

```diff
@@ -27,19 +27,20 @@ bool dep_graph_domaint::merge(
   goto_programt::const_targett from,
   goto_programt::const_targett to)
 {
-  bool changed = false;
+  bool changed = is_bottom() || has_changed;
 
-  if(is_bottom())
+  if(is_bottom() || (has_changed && data_deps.empty()))
   {
-    has_values = tvt::unknown();
     data_deps = src.data_deps;
```

Comment:
Suggest something like the above comment goes into the code as this is entirely non-obvious! 

It also might be clear to do:

```
if(changed)
{
  // ...
  // comment explaining why not if data_deps in non-empty
  if(!data_deps.empty())
  {
    data_deps = src.data_deps;
  }
}
```